pkg/volume/util/util.go

```diff
@@ -270,3 +270,119 @@ func stringToSet(str, delimiter string) (sets.String, error) {
 	}
 	return zonesSet, nil
 }
+
+// MapDevice creates a symbolic link to block device under specified map path
+func MapDevice(devicePath string, mapDir string, linkName string) error {
+	// example of global map path:
+	//   devicePath: plugins/kubernetes.io/{PluginName}/{DefaultKubeletVolumeDevicesDirName}/{volumeName}
+	//   linkName: {podUid}
+	//
+	// example of pod device map path:
+	//   devicePath: pods/{podUid}/{DefaultKubeletVolumeDevicesDirName}/{escapeQualifiedPluginName}
+	//   linkName: {volumeName}
+	if len(devicePath) == 0 {
+		return fmt.Errorf("Failed to map device to map path. devicePath is empty")
+	}
+	if len(mapDir) == 0 {
+		return fmt.Errorf("Failed to map device to map path. mapDir is empty")
+	}
+	if !filepath.IsAbs(mapDir) {
+		return fmt.Errorf("The map path should be absolute: map path: %s", mapDir)
+	}
+	glog.Errorf("#### DEBUG LOG ####: mapDevice devicePath: %s", devicePath)
+	glog.Errorf("#### DEBUG LOG ####: mapDevice mapDir; %s", mapDir)
+
+	_, err := os.Stat(mapDir)
+	if err != nil && !os.IsNotExist(err) {
+		glog.Errorf("cannot validate map path: %s", mapDir)
+		return err
+	}
+	if err = os.MkdirAll(mapDir, 0750); err != nil {
+		return fmt.Errorf("Failed to mkdir %s, error", mapDir)
+	}
+	// If old symlink exits, remove it then create new one
```

Comment:
If it exists, does it mean we didn't clean up correctly?

Reply:
@msau42 
I suppose that there is a case a node accidentally rebooted, and then kubelet recreates symlinks.

If the node is rebooted, order of /dev/sdX may be changed, therefore we should remove old symlink, then create new symlink associated to a new /dev/sdX.

Reply:
How does volume reconstruction fit in here?